File: sdk/src/main/java/com/libreshockwave/format/ScriptFormatUtils.java
package com.libreshockwave.format;

import com.libreshockwave.chunks.ScriptChunk;
import com.libreshockwave.chunks.ScriptNamesChunk;

/**
 * Shared formatting utilities for scripts, literals, and related data.
 * Used by SDK, VM, player, and cast-extractor modules.
 */
public final class ScriptFormatUtils {

    private ScriptFormatUtils() {}

    // ========== Literal Type Formatting ==========

    /**
     * Get a human-readable name for a literal type code.
     * @param typeCode The literal type code (1=string, 4=int, 9=float)
     * @return The type name
     */
    public static String getLiteralTypeName(int typeCode) {
        return switch (typeCode) {
            case 1 -> "string";
            case 4 -> "int";
            case 9 -> "float";
            default -> "type" + typeCode;
        };
    }

    /**
     * Get a short name for a literal type code (for compact display).
     * @param typeCode The literal type code
     * @return The short type name
     */
    public static String getLiteralTypeNameShort(int typeCode) {
        return switch (typeCode) {
            case 1 -> "str";
            case 4 -> "int";
            case 9 -> "float";
            default -> "lit";
        };
    }

    /**
     * Format a literal entry for display.
     * @param lit The literal entry
     * @return Formatted string like "string: \"hello\"" or "int: 42"
     */
    public static String formatLiteral(ScriptChunk.LiteralEntry lit) {
        String typeName = getLiteralTypeName(lit.type());
        String valueStr = formatLiteralValue(lit.value());
        return typeName + ": " + valueStr;
    }

    /**
     * Format a literal value for display.
     * Strings are wrapped in quotes.
     */
    public static String formatLiteralValue(Object value) {
        if (value instanceof String s) {
            return "\"" + s + "\"";
        }
        return String.valueOf(value);
    }

    /**
     * Format a literal value with optional truncation for strings.
     */
    public static String formatLiteralValue(Object value, int maxStringLength) {
        if (value instanceof String s) {
            String truncated = truncate(s, maxStringLength);
            return "\"" + truncated + "\"";
        }
        return String.valueOf(value);
    }

    // ========== Script Type Formatting ==========

    /**
     * Get a human-readable name for a script type.
     * @param scriptType The script type enum
     * @return Display name like "Movie Script", "Behavior", etc.
     */
    public static String getScriptTypeName(ScriptChunk.ScriptType scriptType) {
        if (scriptType == null) return "Unknown";
        return switch (scriptType) {
            case MOVIE_SCRIPT -> "Movie Script";
            case BEHAVIOR -> "Behavior";
            case PARENT -> "Parent Script";
            case SCORE -> "Score Script";
            default -> "Script";
        };
    }

    // ========== Handler/Name Resolution ==========

    /**
     * Resolve a name ID to a string, with fallback if names unavailable.
     * @param names The script names chunk (may be null)
     * @param nameId The name ID to resolve
     * @return The resolved name or a fallback like "#123"
     */
    public static String resolveName(ScriptNamesChunk names, int nameId) {
        if (names != null && nameId >= 0 && nameId < names.names().size()) {
            return names.getName(nameId);
        }
        return "#" + nameId;
    }

    /**
     * Resolve a handler name with a descriptive fallback.
     */
    public static String resolveHandlerName(ScriptNamesChunk names, int nameId) {
        if (names != null && nameId >= 0 && nameId < names.names().size()) {
            return names.getName(nameId);
        }
        return "handler#" + nameId;
    }

    // ========== String Utilities ==========

    /**
     * Truncate a string if it exceeds the maximum length.
     * @param s The string to truncate
     * @param maxLength Maximum length including ellipsis
     * @return The truncated string with "..." if needed
     */
    public static String truncate(String s, int maxLength) {
        if (s == null) return "";
        if (s.length() <= maxLength) return s;
        if (maxLength <= 3) return "...";
        return s.substring(0, maxLength - 3) + "...";
    }

    /**
     * Normalize line endings to spaces (for single-line display).
     */
    public static String normalizeLineEndings(String s) {
        if (s == null) return "";
        return s.replace("\r\n", " ")
                .replace("\r", " ")
                .replace("\n", " ")
                .trim();
    }
}

File: cpp/src/format/ScriptFormatUtils.cpp
#include "libreshockwave/format/ScriptFormatUtils.hpp"

#include "libreshockwave/chunks/ScriptNamesChunk.hpp"
#include "libreshockwave/util/StringUtils.hpp"

namespace libreshockwave::format {
namespace {

std::string literalValueToString(const chunks::ScriptChunk::LiteralValue& value, int maxStringLength, bool truncateStrings) {
    if (std::holds_alternative<std::string>(value)) {
        auto stringValue = std::get<std::string>(value);
        if (truncateStrings) {
            stringValue = truncate(stringValue, maxStringLength);
        }
        return "\"" + stringValue + "\"";
    }
    if (std::holds_alternative<int>(value)) {
        return std::to_string(std::get<int>(value));
    }
    if (std::holds_alternative<std::vector<std::uint8_t>>(value)) {
        return "[bytes:" + std::to_string(std::get<std::vector<std::uint8_t>>(value).size()) + "]";
    }
    return "null";
}

} // namespace

std::string getLiteralTypeName(int typeCode) {
    switch (typeCode) {
        case 1:
            return "string";
        case 4:
            return "int";
        case 9:
            return "float";
        default:
            return "type" + std::to_string(typeCode);
    }
}

std::string getLiteralTypeNameShort(int typeCode) {
    switch (typeCode) {
        case 1:
            return "str";
        case 4:
            return "int";
        case 9:
            return "float";
        default:
            return "lit";
    }
}

std::string formatLiteral(const chunks::ScriptChunk::LiteralEntry& literal) {
    return getLiteralTypeName(literal.type) + ": " + formatLiteralValue(literal.value);
}

std::string formatLiteralValue(const chunks::ScriptChunk::LiteralValue& value) {
    return literalValueToString(value, 0, false);
}

std::string formatLiteralValue(const chunks::ScriptChunk::LiteralValue& value, int maxStringLength) {
    return literalValueToString(value, maxStringLength, true);
}

std::string getScriptTypeName(chunks::ScriptChunkType scriptType) {
    switch (scriptType) {
        case chunks::ScriptChunkType::MovieScript:
            return "Movie Script";
        case chunks::ScriptChunkType::Behavior:
            return "Behavior";
        case chunks::ScriptChunkType::Parent:
            return "Parent Script";
        case chunks::ScriptChunkType::Score:
            return "Score Script";
        case chunks::ScriptChunkType::Unknown:
            return "Script";
    }
    return "Script";
}

std::string resolveName(const chunks::ScriptNamesChunk* names, int nameId) {
    if (names && nameId >= 0 && nameId < static_cast<int>(names->names().size())) {
        return names->getName(nameId);
    }
    return "#" + std::to_string(nameId);
}

std::string resolveHandlerName(const chunks::ScriptNamesChunk* names, int nameId) {
    if (names && nameId >= 0 && nameId < static_cast<int>(names->names().size())) {
        return names->getName(nameId);
    }
    return "handler#" + std::to_string(nameId);
}

std::string truncate(std::string_view value, int maxLength) {
    return util::truncate(value, maxLength);
}

std::string normalizeLineEndings(std::string_view value) {
    std::string normalized;
    normalized.reserve(value.size());
    for (std::size_t index = 0; index < value.size(); ++index) {
        if (value[index] == '\r') {
            normalized.push_back(' ');
            if (index + 1 < value.size() && value[index + 1] == '\n') {
                ++index;
            }
        } else if (value[index] == '\n') {
            normalized.push_back(' ');
        } else {
            normalized.push_back(value[index]);
        }
    }

    const auto start = normalized.find_first_not_of(' ');
    if (start == std::string::npos) {
        return "";
    }
    const auto end = normalized.find_last_not_of(' ');
    return normalized.substr(start, end - start + 1);
}

} // namespace libreshockwave::format
